### PR TITLE
Add support for FETCH clause and FETCH FIRST WITH TIES clause

### PR DESCRIPTION
**Pull request by fgwang7w**
Syntax:

```
<query expression> ::=
  [ <with clause> ] <query expression body>
      [ <order by clause> ] [ <fetch first clause> ]

<fetch first clause> ::=
  FETCH { FIRST | NEXT } [ <fetch first quantity> ] { ROW | ROWS } { ONLY | WITH TIES }

<fetch first quantity> ::=
    <fetch first row count>
  | <fetch first percentage>

<fetch first row count> ::=
  <simple value specification>

<fetch first percentage> ::=
  <simple value specification> PERCENT
```
Extracted-From:  
https://github.com/prestosql/presto/pull/666
https://github.com/prestosql/presto/pull/832

Test plan - Add UT and test locally

```
== RELEASE NOTES ==

General Changes
* Add support for FETCH syntax 
* Add support for FETCH FIRST WITH TIES syntax
```

### CODE REVIEW

**Comment by lucaskim1233** [comment]
Is this feature scheduled for next release?

**Comment by fgwang7w** [comment]
> LGTM

Thanks for reviewing. However, current commit only suppport  FETCH syntax and OFFSET syntax. Add support for FETCH FIRST WITH TIES clause has not yet been committed. OFFSET implementation is still being worked on. 
The code base is quite complicated and direct backporting is not possible which now results in many handy maneuvers. Is it acceptable to include suppport for FETCH syntax and OFFSET syntax only in the next release and another commit in the subsequent release? @kaikalur fyi @rongrong

**Comment by fgwang7w** [comment]
@kaikalur I added the implementation for OFFSET and add support for FETCH FIRST WITH TIES clause. Could you please review again. I'd like to retain 2 commits as the 2nd commit contains the support for FETCH FIRST WITH TIES clause. Thanks.

**Comment by rongrong** [comment]
> > LGTM
> 
> Thanks for reviewing. However, current commit only suppport FETCH syntax and OFFSET syntax. Add support for FETCH FIRST WITH TIES clause has not yet been committed. OFFSET implementation is still being worked on.
> The code base is quite complicated and direct backporting is not possible which now results in many handy maneuvers. Is it acceptable to include suppport for FETCH syntax and OFFSET syntax only in the next release and another commit in the subsequent release? @kaikalur fyi @rongrong

I'm OK with leaving `FETCH FIRST WITH TIES` into a different PR if it's quite complicated. Please update the description accordingly if you decided to do so. The test failures seem to be related. Please fix them. Thanks!

**Comment by fgwang7w** [comment]
> > > LGTM
> > 
> > 
> > Thanks for reviewing. However, current commit only suppport FETCH syntax and OFFSET syntax. Add support for FETCH FIRST WITH TIES clause has not yet been committed. OFFSET implementation is still being worked on.
> > The code base is quite complicated and direct backporting is not possible which now results in many handy maneuvers. Is it acceptable to include suppport for FETCH syntax and OFFSET syntax only in the next release and another commit in the subsequent release? @kaikalur fyi @rongrong
> 
> I'm OK with leaving `FETCH FIRST WITH TIES` into a different PR if it's quite complicated. Please update the description accordingly if you decided to do so. The test failures seem to be related. Please fix them. Thanks!

Acutually `FETCH FIRST TIES` support is included in the same PR now. I am working on fixing the test failures now.

**Comment by fgwang7w** [comment]
All tests fixed. Please help review the code for next release. Thanks!

**Comment by fgwang7w** [comment]
@rongrong could you please see this mr is ready to merge? I see another reviewer marked LGTM. thank you!

**Comment by kaikalur** [comment]
This PR is too big now. You should try and split it into several smaller self-contained ones and/or add session param(s) to turn the feature(s) off. I haven't looked to see if you already did that - the PR is just too big for me to review in the broken github UI lol

**Comment by fgwang7w** [comment]
> This PR is too big now. You should try and split it into several smaller self-contained ones and/or add session param(s) to turn the feature(s) off. I haven't looked to see if you already did that - the PR is just too big for me to review in the broken github UI lol

Thanks for the comment, are you suggesting to split the code into multiple commits? Refactoring the code might take a very long time and error-prune, since I have done unit-tests and running test locally, will it be split the review task at component-level? I will keep in mind to reduce the code size per commit for future contributions

**Comment by tdcmeehan** [comment]
I agree with @kaikalur, let's split this into parts.

**Comment by fgwang7w** [comment]
@kaikalur @tdcmeehan @rongrong  I have reworked this PR to 3 multiple commits to provide a reviewable format. Given the size of the code and multiple file intersections, I will keep support for FETCH FIRST WITH TIES on a separate commit by another PR later. Please help review this PR for FETCH and OFFSET support only first. Thank you!

**Comment by fgwang7w** [comment]
Hi @tdcmeehan, this PR is ready for review. I splitted the code into 4 commits now. many thanks!

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 1]
```diff
@@ -886,6 +916,51 @@ public void testSelectWithOrderBy()
...
+                                Optional.empty(),
+                                Optional.of(new FetchFirst(Optional.empty()))),
+                        Optional.empty(),
+                        Optional.empty()));
+
+        Query valuesQuery = query(values(
+                row(new LongLiteral("1"), new StringLiteral("1")),
+                row(new LongLiteral("2"), new StringLiteral("2"))));
+
+        assertStatement("SELECT * FROM (VALUES (1, '1'), (2, '2')) FETCH FIRST ROW ONLY",
```
Comment (tdcmeehan) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:1012]: Can we add some more tests for invalid inputs?

**Review by tdcmeehan — COMMENTED** [review]
`Add support for FETCH syntax` ❓

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Offset.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    public Offset(NodeLocation location, String rowCount)
+    {
+        this(Optional.of(location), rowCount);
+    }
+
+    public Offset(Optional<NodeLocation> location, String rowCount)
+    {
+        super(location);
+        this.rowCount = rowCount;
```
Comment (tdcmeehan) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Offset.java:42]:
```suggestion
        this.rowCount = requireNonNull(rowCount, "rowCount is null");
```

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 3]
```diff
@@ -911,8 +934,39 @@ public void testSelectWithOrderBy()
...
                                 Optional.empty()),
                         Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty()));
+    }
+
+    @Test
+    public void testSelectWithOffset()
+    {
+        assertStatement("SELECT * FROM table1 OFFSET 2 ROWS",
```
Comment (tdcmeehan) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:947]: Similar as elsewhere, could we verify a error when the OFFSET is specified incompletely?  Is it correct to allow a number > 1 that is specified with `ROW`?

File: presto-verifier/src/test/java/com/facebook/presto/verifier/framework/TestLimitQueryDeterminismAnalyzer.java [hunk 4]
```diff
@@ -128,8 +128,8 @@ public void testNotRunLimitNoOrderBy()
         assertAnalysis(prestoAction, "INSERT INTO test SELECT * FROM (SELECT * FROM source LIMIT 1000)", NOT_RUN);
 
         // LIMIT ALL clause
-        assertAnalysis(prestoAction, "INSERT INTO test SELECT * FROM source LIMIT all", NOT_RUN);
-        assertAnalysis(prestoAction, "CREATE TABLE test AS (WITH f AS (select * from g) ((SELECT * FROM source UNION ALL SELECT * FROM source LIMIT ALL)))", NOT_RUN);
+        assertAnalysis(prestoAction, "INSERT INTO test SELECT * FROM source LIMIT all", DETERMINISTIC);
```
Comment (tdcmeehan) [inline presto-verifier/src/test/java/com/facebook/presto/verifier/framework/TestLimitQueryDeterminismAnalyzer.java:131]: How is this commit related?

**Review by tdcmeehan — COMMENTED** [review]
`Add support for OFFSET syntax` ❓

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/FetchFirst.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    public FetchFirst(Optional<String> rowCount, boolean withTies)
+    {
+        this(Optional.empty(), rowCount, withTies);
+    }
+
+    public FetchFirst(Optional<NodeLocation> location, Optional<String> rowCount, boolean withTies)
+    {
+        super(location);
+        this.rowCount = rowCount;
+        this.withTies = withTies;
```
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/FetchFirst.java:54]: it's possible to have empty rowCount when sql statement is FETCH NEXT, refer to testSelectWithFetch

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java (hunk 1 above)
Comment (fgwang7w) [reply presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:1033]: added more UTs for the invalid input in testAnalayzer to fail on statement analyzer phase

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java (hunk 3 above)
Comment (fgwang7w) [reply presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:947]: yes I add more UT checks on analyzer to catch and verify

File: presto-verifier/src/test/java/com/facebook/presto/verifier/framework/TestLimitQueryDeterminismAnalyzer.java (hunk 4 above)
Comment (fgwang7w) [reply presto-verifier/src/test/java/com/facebook/presto/verifier/framework/TestLimitQueryDeterminismAnalyzer.java:129]: LIMIT all is of DETERMINISTIC query

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 6]
```diff
@@ -2253,18 +2296,18 @@ public void testLogicalExplainJsonFormat()
...
         assertEquals(getOnlyElement(result.getOnlyColumnAsSet()), getExplainPlan(query, DISTRIBUTED));
     }
 
     @Test
-    public void testDistributedExplainTextFormat()
+    public void testDistributedExplain()
     {
         String query = "SELECT * FROM orders";
-        MaterializedResult result = computeActual("EXPLAIN (TYPE DISTRIBUTED, FORMAT TEXT) " + query);
+        MaterializedResult result = computeActual("EXPLAIN (TYPE DISTRIBUTED) " + query);
```
Comment (tdcmeehan) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:2335]: Unrelated?

**Review by tdcmeehan — COMMENTED** [review]
`Add OFFSET implementation` ✅

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -2254,7 +2264,10 @@ private boolean analyzeLimit(FetchFirst node)
                 analysis.setLimit(node, rowCount);
             }
 
-            return node.isWithTies();
+            if (node.isWithTies()) {
+                return true;
+            }
+            return false;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2270]: Is this necessary?

**Review by tdcmeehan — COMMENTED** [review]
`Add support for FETCH FIRST WITH TIES` ✅

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
`add more ut` ❌   Could you squash this commit with the appropriate previous commits?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 6 above)
Comment (fgwang7w) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:2335]: yah not related, removed, thanks

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 7 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2270]: this is part of implementation for Fetch first with Ties that needs to analylze if sementically correct when limit node has Ties presented, needed

**Comment by fgwang7w** [comment]
> `add more ut` ❌ Could you squash this commit with the appropriate previous commits?

done, please review my feedback and any other revision needed. many thanks

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 7 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2270]: I meant, could we simply `return node.isWithTies()`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 8]
```diff
@@ -611,6 +612,20 @@ public PlanOptimizers(
         this.runtimeOptimizers = runtimeBuilder.build();
     }
 
+    @PostConstruct
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:615]: Unrelated change--please revert

**Comment by tdcmeehan** [comment]
@kaikalur would you like to give this a review now?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 8 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 7 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: righty, fixed

**Review by fgwang7w — COMMENTED** [review]
reverted

**Comment by aweisberg** [comment]
Was there code in this PR copied from another project? If so please follow the [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).

**Comment by aweisberg** [comment]
Just FYI it seems like 3/4 of the commits don't have a co-author tag. Thanks!

**Comment by fgwang7w** [comment]
> Just FYI it seems like 3/4 of the commits don't have a co-author tag. Thanks!

commit msg is fixed

**Comment by fgwang7w** [comment]
@kaikalur had an LGTM on Jul 23, please review again. @tdcmeehan please help move forward for this PR. this has been waiting for too long to get merged since  Jul 23

**Comment by kaikalur** [comment]
So I'm still confused - can we create 4 separate PRs? That'll be easier to review/track.

**Comment by fgwang7w** [comment]
> So I'm still confused - can we create 4 separate PRs? That'll be easier to review/track.

I think it's already splitted into 4 commits which can be reviewed individually as Tim did.

**Comment by kaikalur** [comment]
> > So I'm still confused - can we create 4 separate PRs? That'll be easier to review/track.
> 
> I think it's already splitted into 4 commits which can be reviewed individually as Tim did. are you suggesting to split those 4 commits into 4 PRs now?

I think that will be better.

**Comment by kaikalur** [comment]
> > > So I'm still confused - can we create 4 separate PRs? That'll be easier to review/track.
> > 
> > 
> > I think it's already splitted into 4 commits which can be reviewed individually as Tim did. are you suggesting to split those 4 commits into 4 PRs now?
> 
> I think that will be better.

So chatted with Rongrong offline and it's ok to have just one PR with 4 commits. I will review them before Monday.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 9]
```diff
@@ -2182,6 +2194,64 @@ private void analyzeOrderBy(Query node, Scope orderByScope)
...
+        /**
+         * @return true if the Query / QuerySpecification containing the analyzed
+         * Limit or FetchFirst, must contain orderBy (i.e., for FetchFirst with ties).
+         */
+        private boolean analyzeLimit(Node node)
+        {
+            checkState(
+                    node instanceof FetchFirst || node instanceof Limit,
+                    "Invalid limit node type. Expected: FetchFirst or Limit. Actual: %s", node.getClass().getName());
+            if (node instanceof FetchFirst) {
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2240]: Can't you just visit the node?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 10]
```diff
@@ -2182,6 +2194,64 @@ private void analyzeOrderBy(Query node, Scope orderByScope)
...
+            }
+            else {
+                return analyzeLimit((Limit) node);
+            }
+        }
+
+        private boolean analyzeLimit(FetchFirst node)
+        {
+            if (!node.getRowCount().isPresent()) {
+                analysis.setLimit(node, 1);
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2251]: When does this happen?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 11]
```diff
@@ -2182,6 +2194,64 @@ private void analyzeOrderBy(Query node, Scope orderByScope)
...
+                analysis.setLimit(node, OptionalLong.empty());
+            }
+            else {
+                long rowCount;
+                try {
+                    rowCount = Long.parseLong(node.getLimit());
+                }
+                catch (NumberFormatException e) {
+                    throw new SemanticException(INVALID_LIMIT_ROW_COUNT, node, "Invalid LIMIT row count: %s", node.getLimit());
+                }
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2282]: Negative is OK here?

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeInputRewrite.java [hunk 12]
```diff
@@ -113,10 +114,10 @@ protected Node visitDescribeInput(DescribeInput node, Void context)
 
             // return the positions and types of all parameters
             Row[] rows = parameters.stream().map(parameter -> createDescribeInputRow(parameter, analysis)).toArray(Row[]::new);
-            Optional<String> limit = Optional.empty();
+            Optional<Node> limit = Optional.empty();
             if (rows.length == 0) {
                 rows = new Row[] {row(new NullLiteral(), new NullLiteral())};
-                limit = Optional.of("0");
+                limit = Optional.of(new Limit("0"));
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeInputRewrite.java:120]: I thought you were making it 1 in the StatementAnalyzer?

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 13]
```diff
@@ -337,6 +339,19 @@ public void testOrderByNonComparable()
...
+    public void testFetchFirstInvalidRowCount()
+    {
+        assertFails(INVALID_FETCH_FIRST_ROW_COUNT, "SELECT * FROM t1 FETCH FIRST 987654321098765432109876543210 ROWS ONLY");
+        assertFails(INVALID_FETCH_FIRST_ROW_COUNT, "SELECT * FROM t1 FETCH FIRST 0 ROWS ONLY");
+    }
+
+    @Test
+    public void testLimitInvalidRowCount()
+    {
+        assertFails(INVALID_LIMIT_ROW_COUNT, "SELECT * FROM t1 LIMIT 987654321098765432109876543210");
```
Comment (kaikalur) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:368]: Add test for negative number as well for more coverage

**Review by kaikalur — CHANGES_REQUESTED** [review]
For the fist commit.

**Comment by NikhilCollooru** [comment]
Hi @fgwang7w can you please address the comments and update the PR ?

We need OFFSET feature and hence we need this PR . Can you please fix the PR ?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 10 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2251]: `FETCH FIRST ROWS` without specifying actual number of rows is allowed, meaning the literal for fetch first quantity can be omitted, in which case it defaults to 1

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 9 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2240]: this was from original cherrypick, but visitor approach makes more sense, fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 11 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:2282]: no the parser only allows digit+ so numeric literal only, `-` is considered as an invalid symbol, here is testing if the literal is not parseable e.g. numeric out of bound

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeInputRewrite.java (hunk 12 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeInputRewrite.java:120]: that was when the `fetch first` was provided, this is for the default limit

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java (hunk 13 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:368]: negative number will be catched by the parser, added ut `testSelectFetchFirstException` to cover the logic path

**Comment by fgwang7w** [comment]
> Hi @fgwang7w can you please address the comments and update the PR ?
> 
> We need OFFSET feature and hence we need this PR . Can you please fix the PR ?

Supoprt for OFFSET clause is now implemented via this PR https://github.com/prestodb/presto/pull/16035

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 14]
```diff
@@ -334,7 +335,8 @@ public PlanOptimizers(
                         estimatedExchangesCostCalculator,
                         ImmutableSet.of(
                                 new ImplementBernoulliSampleAsFilter(),
-                                new ImplementOffset())),
+                                new ImplementOffset(),
+                                new ImplementLimitWithTies(metadata))),
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:339]: We need this to be controlled by a session param/feature config. In fact, I would like the whole feature to be optional.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 15]
```diff
@@ -887,29 +890,17 @@ private PlanBuilder distinct(PlanBuilder subPlan, QuerySpecification node)
...
-    {
-        return sort(subPlan, node.getOrderBy(), analysis.getOrderByExpressions(node));
-    }
-
-    private PlanBuilder sort(PlanBuilder subPlan, Optional<OrderBy> orderBy, List<Expression> orderByExpressions)
+    private Optional<OrderingScheme> orderingScheme(PlanBuilder subPlan, Optional<OrderBy> orderBy, List<Expression> orderByExpressions)
     {
         if (!orderBy.isPresent() || (isSkipRedundantSort(session)) && analysis.isOrderByRedundant(orderBy.get())) {
-            return subPlan;
+            return Optional.empty();
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java:896]: Why do we need this change?

**Review by kaikalur — CHANGES_REQUESTED** [review]
I put a couple of comments. But my high level comment would be to keep this PR non-intrusive. I see a lot of core planner changes. I'm not sure why those are needed.
